src/model/enhance-images.ts

```diff
@@ -0,0 +1,337 @@
+					caption: extractCaption(nextElement),
+				},
+			} as ImageBlockElement);
+			// Remove the next element
+			elements.splice(i + 1, 1);
+		} else if (
+			isImage(thisElement) &&
+			isTitle(nextElement) &&
+			isCaption(subsequentElement)
+		) {
+			const thisImage = thisElement as ImageBlockElement;
+			withSpecialCaptions.push({
+				...thisImage,
+				data: {
+					...thisImage.data,
+					caption: extractCaption(subsequentElement),
+				},
+			} as ImageBlockElement);
+			// Remove the subsequent element
+			elements.splice(i + 2, 1);
+		} else {
+			// Pass through
+			withSpecialCaptions.push(thisElement);
+		}
+	});
+	return withSpecialCaptions;
+};
+
+const addCaptionsToMultis = (elements: CAPIElement[]): CAPIElement[] => {
+	const withSpecialCaptions: CAPIElement[] = [];
+	elements.forEach((thisElement, i) => {
+		const nextElement = elements[i + 1];
+		const subsequentElement = elements[i + 2];
+		if (isMultiImage(thisElement) && isCaption(nextElement)) {
+			withSpecialCaptions.push({
+				...thisElement,
+				caption: extractCaption(nextElement),
+			} as MultiImageBlockElement);
+			// Remove the next element
+			elements.splice(i + 1, 1);
+		} else if (
+			isMultiImage(thisElement) &&
+			isTitle(nextElement) &&
+			isCaption(subsequentElement)
+		) {
+			withSpecialCaptions.push({
+				...thisElement,
+				caption: extractCaption(subsequentElement),
+			} as MultiImageBlockElement);
+			// Remove the subsequent element
+			elements.splice(i + 2, 1);
+		} else {
+			// Pass through
+			withSpecialCaptions.push(thisElement);
+		}
+	});
+	return withSpecialCaptions;
+};
+
+const stripCaptions = (elements: CAPIElement[]): CAPIElement[] => {
+	// Remove all captions from all images
+	const withoutCaptions: CAPIElement[] = [];
+	elements.forEach((thisElement) => {
+		if (
+			thisElement._type ===
+			'model.dotcomrendering.pageElements.ImageBlockElement'
+		) {
+			// Remove the caption from this image
+			withoutCaptions.push({
+				...thisElement,
+				data: {
+					...thisElement.data,
+					caption: '',
+				},
+			} as ImageBlockElement);
+		} else {
+			// Pass through
+			withoutCaptions.push(thisElement);
+		}
+	});
+	return withoutCaptions;
+};
+
+class Enhancer {
+	elements: CAPIElement[];
+
+	constructor(elements: CAPIElement[]) {
+		this.elements = elements;
+	}
+
+	stripCaptions() {
+		this.elements = stripCaptions(this.elements);
+		return this;
+	}
+
+	addMultiImageElements() {
+		this.elements = addMultiImageElements(this.elements);
+		return this;
+	}
+
+	addTitles() {
+		this.elements = addTitles(this.elements);
+		return this;
+	}
+
+	addCaptionsToMultis() {
+		this.elements = addCaptionsToMultis(this.elements);
+		return this;
+	}
+
+	addCaptionsToImages() {
+		this.elements = addCaptionsToImages(this.elements);
+		return this;
+	}
+}
+
+const enhance = (
+	elements: CAPIElement[],
+	isPhotoEssay: boolean,
+): CAPIElement[] => {
+	if (isPhotoEssay) {
+		return (
+			new Enhancer(elements)
```

Comment:
The Enhancer class is probably the only bit here I'm less keen on. It adds a load of boilerplate to wrap the functions for little value. Controversial proposal, as know `reduce` can be hard to read, but what about, e.g.:

```
const elems = [1, 2, 3];
const double = (ns: number[]): number[] => ns.map((n) => n * 2);
const half = (ns: number[]): number[] => ns.map((n) => n / 2);
[double, half].reduce((acc, fn) => fn(acc), elems);
```

(I used `number[]` here for simplicity but imagine `CAPIElement[]` instead.)